sympy/solvers/diophantine/diophantine.py

```diff
@@ -1349,30 +1349,47 @@ def diophantine(eq, param=symbols("t", integer=True), syms=None,
     if isinstance(eq, Eq):
         eq = eq.lhs - eq.rhs
 
+    var = list(eq.expand(force=True).free_symbols)
+    var.sort(key=default_sort_key)
+    if syms:
+        if not is_sequence(syms):
+            raise TypeError(
+                'syms should be given as a sequence, e.g. a list')
+        syms = [i for i in syms if i in var]
+        if syms != var:
+            dict_sym_index = dict(zip(syms, range(len(syms))))
+            return {tuple([t[dict_sym_index[i]] for i in var])
+                        for t in diophantine(eq, param, permute=True)}
+    n, d = eq.as_numer_denom()
+    if n.is_number:
+        return set()
+    if not d.is_number:
+        dsol = diophantine(d)
+        good = diophantine(n) - dsol
+        return {s for s in good if _mexpand(d.subs(zip(var, s)))}
+    else:
+        eq = n
+    eq = factor_terms(eq)
+    assert not eq.is_number
+    eq = eq.as_independent(*var, as_Add=False)[1]
+
+    # check for plausible integer non-solutions
+    ivar = {s: Dummy(integer=True) for s in var if not s.is_integer}
```

Comment:
I think we should define private methods for these regular checks, to make code readable and reusable.